mio/src/lib.rs

```diff
@@ -198,39 +283,15 @@ impl Loop {
                          mio::PollOpt::edge()));
         Ok(Loop {
             io: io,
-            done: HashMap::new(),
-            next: 1,
             tx: tx,
             rx: rx,
+            dispatch: HashMap::default(),
+            token_counter: TokenCounter::new(),
         })
     }
 
-    pub fn await<F: Future>(&mut self, mut f: F)
-                            -> Result<F::Item, F::Error> {
-        let (tx, rx) = channel();
-        f.schedule(move |r| {
-            drop(tx.send(r))
-            // TODO: signal to the event loop that it should wake up
-        });
-        let mut ret = None;
-        self._await(&mut || {
-            match rx.try_recv() {
-                Ok(e) => ret = Some(e),
-                Err(TryRecvError::Empty) => {}
-                Err(TryRecvError::Disconnected) => panic!(),
-            }
-            ret.is_some()
-        });
-        match ret.unwrap() {
-            Ok(e) => Ok(e),
-            Err(PollError::Other(e)) => Err(e),
-            Err(PollError::Panicked(p)) => panic::resume_unwind(p),
-            Err(PollError::Canceled) => panic!("canceled"),
-        }
-    }
-
-    fn _await(&mut self, done: &mut FnMut() -> bool) {
-        while !done() {
+    pub fn run(&mut self) {
```

Comment:
It may be worth getting `cargo test` to work in this directory perhaps to see how this hooks up with driving a future